### ISSUE TITLE
Work out whether any changes are needed for positional/keyword args in Ruby 2.7

### ISSUE THREAD

**Issue by floehopper** [freerange/mocha#445]
See: https://www.ruby-lang.org/en/news/2019/12/12/separation-of-positional-and-keyword-arguments-in-ruby-3-0/

**Comment by stripedpumpkin** [comment]
Getting many warnings with ruby 2.7.0 and mocha 1.11.2 along the same lines, here is a simple example below.

```ruby
# frozen_string_literal: true

require 'test_helper'

class MochaTest < ActiveSupport::TestCase
  class DuckFeeder
    def initialize(duck:)
      @duck = duck
    end

    def call
      duck.feed(
        food: 'snails',
        amount_in_g: 100
      )
    end

    private

    attr_reader :duck
  end

  class FakeDuck
    def feed(**kwargs)
      quack "Delicious #{kwargs.fetch(:food)}!"
    end

    private

    def quack(something)
      puts something
    end
  end

  test 'duck was fed' do
    duck = FakeDuck.new
    # Line 37 is the one below
    duck.expects(:feed).with do |**kwargs|
      assert_equal(
        { food: 'snails', amount_in_g: 100 },
        kwargs
      )
    end
    DuckFeeder.new(duck: duck).call
  end
end
```
Gives the following warning
=> 
```ruby/gems/2.7.0/gems/mocha-1.11.2/lib/mocha/parameters_matcher.rb:13: warning: Using the last argument as keyword parameters is deprecated; maybe ** should be added to the call```
```test/poros/mocha_test.rb:37: warning: The called method `call' is defined here```

(In this previous case, we do not have to use a block. But in some cases using the block is really needed. For example, when creating an ActiveRecord, its `id` might not be known in advance, and we have to use the block structure to check *afterwards* that ids match for example.)

```ruby
  test 'fresh duck is getting assigned to a feeder' do
    DuckFeederAssigner.expects(:new).with do |**kwargs|
      @actual_duck_id = kwargs.fetch(:duck_id)
      true
    end.once
    bob = FakeDuck.create! name: 'Bob'
    assert_equal(
      @actual_duck_id,
      bob.id
    )
  end
```

**Comment by floehopper** [comment]
@stripedpumpkin Thanks for reporting this. We'll investigate as soon as we have time.

**Comment by floehopper** [comment]
@stripedpumpkin 

At first glance, I'm not sure there's a _simple_ way to provide a _complete_ fix for your scenario. However, I think the following might be a reasonable workaround. If you change the expectation in your first example to use `kwargs` vs `**kwargs`, then `kwargs` will be a `Hash` which you can use in your assertion:

```ruby
    duck.expects(:feed).with do |kwargs|
      assert_equal(
        { food: 'snails', amount_in_g: 100 },
        kwargs
      )
    end
```

Let me know what do you think?

**Comment by stripedpumpkin** [comment]
@floehopper Thanks for the suggestion. I think it would do it. Even with `*args, **kwargs` type scenarios. I'll let you know if I find a case that cannot be fixed this way.

Still at some point it would be nice to keep the `**kwargs` if only because it is more expressive.

**Comment by floehopper** [comment]
I'm hoping this is addressed by #554. Closing.
